pkg/buildid/note.go

```diff
@@ -38,19 +39,37 @@ import (
 
 const (
 	maxNoteSize        = 1 << 20 // in bytes
-	NoteTypeGNUBuildID = 3
-	NoteTypeGoBuildID  = 4
+	noteTypeGNUBuildID = 3
+	noteTypeGoBuildID  = 4
 )
 
-// ElfNote is the payload of a Note Section in an ELF file.
-type ElfNote struct {
+// elfNote is the payload of a Note Section in an ELF file.
+type elfNote struct {
 	Name string // Contents of the "name" field, omitting the trailing zero byte.
 	Desc []byte // Contents of the "desc" field.
 	Type uint32 // Contents of the "type" field.
 }
 
-// ParseNotes returns the notes from a SHT_NOTE section or PT_NOTE segment.
-func ParseNotes(reader io.Reader, alignment int, order binary.ByteOrder) ([]ElfNote, error) {
+// findInNotes finds the build id in the given ELF file's notes section.
+func findInNotes(ef *elf.File, section string, predicate func(notes []elfNote) ([]byte, error)) ([]byte, error) {
```

Comment:
I find this pattern with the closure unnecessarily complicated, why not just return the notes and in each spot where it's used then run the function explicitly.

Each of the closures are only used once, and are executed at the very end of this function.

Reply:
Makes sense. I will refactor.